spring-kafka/src/main/java/org/springframework/kafka/support/serializer/JsonDeserializer.java

```diff
@@ -91,29 +99,74 @@
 
 	private boolean typeMapperExplicitlySet = false;
 
+	/**
+	 * Construct an instance with a default {@link ObjectMapper}.
+	 */
 	public JsonDeserializer() {
-		this((Class<T>) null);
+		this((Class<T>) null, true);
 	}
 
-	protected JsonDeserializer(ObjectMapper objectMapper) {
-		this(null, objectMapper);
+	/**
+	 * Construct an instance with the provided {@link ObjectMapper}.
+	 * @param objectMapper a custom object mapper.
+	 */
+	public JsonDeserializer(ObjectMapper objectMapper) {
+		this(null, objectMapper, true);
 	}
 
+	/**
+	 * Construct an instance with the provided target type, and a default
+	 * {@link ObjectMapper}.
+	 * @param targetType the target type to use if no type info headers are present.
+	 */
 	public JsonDeserializer(Class<T> targetType) {
-		this(targetType, new ObjectMapper());
+		this(targetType, true);
+	}
+
+	/**
+	 * Construct an instance with the provided target type, and
+	 * useHeadersIfPresent with a default {@link ObjectMapper}.
+	 * @param targetType the target type.
+	 * @param useHeadersIfPresent true to use headers if present and fall back to target
+	 * type if not.
+	 * @since 2.2
+	 */
+	public JsonDeserializer(Class<T> targetType, boolean useHeadersIfPresent) {
+		this(targetType, new ObjectMapper(), useHeadersIfPresent);
 		this.objectMapper.configure(MapperFeature.DEFAULT_VIEW_INCLUSION, false);
 		this.objectMapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
 	}
 
-	@SuppressWarnings("unchecked")
+	/**
+	 * Construct an instance with the provided target type, and {@link ObjectMapper}.
+	 * @param targetType the target type to use if no type info headers are present.
+	 * @param objectMapper the mapper. type if not.
+	 */
 	public JsonDeserializer(Class<T> targetType, ObjectMapper objectMapper) {
+		this(targetType, objectMapper, true);
+	}
+
+	/**
+	 * Construct an instance with the provided target type, {@link ObjectMapper} and
+	 * useHeadersIfPresent.
+	 * @param targetType the target type.
+	 * @param objectMapper the mapper.
+	 * @param useHeadersIfPresent true to use headers if present and fall back to target
+	 * type if not.
+	 * @since 2.2
+	 */
+	@SuppressWarnings("unchecked")
+	public JsonDeserializer(@Nullable Class<T> targetType, ObjectMapper objectMapper, boolean useHeadersIfPresent) {
 		Assert.notNull(objectMapper, "'objectMapper' must not be null.");
+		Assert.isTrue(targetType != null || useHeadersIfPresent,
```

Comment:
This one has to go after `resolveGeneric(0)` a bit bellow.
It can be `null` here, but we may resolve a target type from the generics.

WDYT?

Reply:
Good catch